### PR TITLE
Add support for SYMM level 3 operator

### PR DESCRIPTION
**Pull request by aacostadiaz**
This PR adds support for the SYMM operator. It's implemented on top of the current GEMM operator by extending the gemm_local implementation.

### CODE REVIEW

File: cmake/CmakeFunctionHelper.cmake [hunk 1]
```diff
@@ -519,71 +519,77 @@ function(add_gemm_configuration
...
-              WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
-              VERBATIM
-            )
-            list(APPEND gemm_sources "${LOCATION}/${file_name}")
-            set(gemm_sources "${gemm_sources}" PARENT_SCOPE)
-          endforeach(index)
-      endforeach(is_beta_zero)
-    endforeach(trans_b)
-  endforeach(trans_a)
+  foreach(symm_a ${boolean_list})
```
Comment (aacostadiaz) [inline cmake/CmakeFunctionHelper.cmake:522]: Mostly esthetic changes. New code is in these two lines and in lines 578-579

File: samples/CMakeLists.txt [hunk 2]
```diff
@@ -22,6 +23,9 @@ foreach(src_file ${SAMPLES_LIST})
   else()
     set_target_properties(${sample_exec} PROPERTIES CXX_STANDARD 14)
   endif()
+  target_compile_definitions(${sample_exec} PRIVATE -D${TUNING_TARGET})
+  target_compile_definitions(${sample_exec} PUBLIC GEMM_VECTORIZATION_SUPPORT=1)
+
```
Comment (aacostadiaz) [inline samples/CMakeLists.txt:28]: Needed to select the appropriate backend and to enable vectorization on the sample app.

File: src/interface/symm_interface.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+  _side = tolower(_side);
+  _uplo = tolower(_uplo);
+  
+  if (_uplo != UPLO_UPPER && _uplo != UPLO_LOWER) {
+    throw std::invalid_argument("invalid _uplo");
+  }
+  if (_side == SIDE_LEFT) { // C <- alpha * A * B + beta * C
+    char trans_symm = _uplo == UPLO_UPPER ? TRANS_YES : TRANS_NO;
+    return _gemm_backend<true, false>(sb_handle, trans_symm, TRANS_NO,
```
Comment (aacostadiaz) [inline src/interface/symm_interface.hpp:58]: SYMM is basically a GEMM wrapper + information about symmetric matrices.

File: test/unittest/blas3/blas3_gemm_common.hpp [hunk 4]
```diff
@@ -19,7 +19,7 @@
  *
  *  SYCL-BLAS: BLAS implementation using SYCL
  *
- *  @filename blas3_gemm_values.hpp
+ *  @filename blas3_gemm_common.hpp
```
Comment (aacostadiaz) [inline test/unittest/blas3/blas3_gemm_common.hpp:22]: TOL: style checks should be able to check for this kind of thing.

File: src/operations/blas3/gemm_local.hpp [hunk 5]
```diff
@@ -542,27 +582,42 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, TileType,
...
+        item_id, row_a, col_a, A, lda, sA,
         [&](index_t, index_t cr) SYCL_BLAS_ALWAYS_INLINE { return cr < m; },
         [&](index_t ic, index_t cc)
-            SYCL_BLAS_ALWAYS_INLINE { return cc < k - ic; });
+            SYCL_BLAS_ALWAYS_INLINE { return cc < k - ic; },
+        [&](index_t row, index_t col) SYCL_BLAS_ALWAYS_INLINE {
+          // If A is symmetric, the valid values are expected on the lower
+          // triangle unless A is transposed, in which case valid values will
+          // be on the upper side.
+          return symm_a ? trans_a ? row <= col : row >= col : true;
```
Comment (aacostadiaz) [inline src/operations/blas3/gemm_local.hpp:607]: Whether `row`, `col` are in the valid `side/uplo` on symmetric matrices.

File: src/operations/blas3/gemm_local.hpp [hunk 6]
```diff
@@ -739,6 +816,54 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, TileType,
...
+      packetize_t::template load<trans, internal, lds>(
+        in_range, src, dest, edge_in_range
+      );
+    } else {
+      vector_t packet{};
+      for (index_t i = 0; i < packet_size; ++i) {
+        reinterpret_cast<value_t *>(&packet)[i] =
+          edge_in_range(i) ? in_triangle(row + i, curr_col) ? 
+            *(src + i) :
+            *((src + i) + ((row + i) - curr_col) * ld + (curr_col - (row + i))) :
```
Comment (aacostadiaz) [inline src/operations/blas3/gemm_local.hpp:844]: If the value is outside of the valid triangle, swap `row` and `col`

File: src/operations/blas3/gemm_local.hpp [hunk 7]
```diff
@@ -739,6 +816,54 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, TileType,
...
+            typename SrcPointerType, typename DestPointerType,
+            typename EdgePredicate, typename TrianglePredicate>
+  static SYCL_BLAS_INLINE typename std::enable_if<!internal>::type load_symm(
+      index_t row , index_t col, index_t col_ofs, index_t ld,
+      const bool in_range, SrcPointerType src, DestPointerType dest,
+      EdgePredicate edge_in_range, TrianglePredicate in_triangle) {
+    const index_t curr_col = col + col_ofs;
+    auto ptr = src;
+    if (!in_triangle(row, curr_col)) {
+      ptr += (row - curr_col) * ld + (curr_col - row);
```
Comment (aacostadiaz) [inline src/operations/blas3/gemm_local.hpp:860]: Swap `row` and `col` to stay in the valid triangle.

File: src/interface/blas3/backend/amd_gpu.hpp [hunk 8]
```diff
@@ -30,8 +30,8 @@ namespace blas {
 namespace gemm {
 
 namespace backend {
-template <bool _t_a, bool _t_b, bool is_beta_zero, typename sb_handle_t,
-          typename container_0_t, typename container_1_t,
+template <bool _t_a, bool _t_b, bool s_a, bool s_b, bool is_beta_zero,
```
Comment (aacostadiaz) [inline src/interface/blas3/backend/amd_gpu.hpp:33]: This is just add `s_a` and `s_b` to all the template `Gemm_Launcher` templates on the backend folder.

File: src/interface/blas3/backend/intel_gpu.hpp [hunk 9]
```diff
@@ -173,9 +173,9 @@ typename sb_handle_t::event_t _gemm(
                                                                 _beta, _c, _ldc,
                                                                 _stridec,
                                                                 batch_size);
-  } else if (_t_b && !_t_a) {
+  } else if (_t_b && !_t_a && !s_a && !s_b) {
```
Comment (aacostadiaz) [inline src/interface/blas3/backend/intel_gpu.hpp:176]: symm is not implemented in `gemm_memory_t::no_local`.

File: benchmark/syclblas/blas3/symm.cpp [hunk 10]
```diff
@@ -0,0 +1,163 @@
+/**************************************************************************
+ *
+ *  @license
+ *  Copyright (C) 2016 Codeplay Software Limited
```
Comment (muhammad-tanvir-1211) [inline benchmark/syclblas/blas3/symm.cpp:4]: Could you please remove the year from this line?

File: benchmark/syclblas/blas3/symm.cpp [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+    double mem_readA = k_d * (k_d + 1) / 2;
+    double mem_readB = m_d * n_d;
+    double mem_writeC = m_d * n_d;
+    double mem_readC = (beta != scalar_t{0}) ? m_d * n_d : 0;
+    double total_mem =
+        (mem_readA + mem_readB + mem_readC + mem_writeC) * sizeof(scalar_t);
+    state.counters["bytes_processed"] = total_mem;
+    state.SetBytesProcessed(state.iterations() * total_mem);
+
+    double nflops_AtimesB = (2 * k_d - 1) * m_d * n_d;
```
Comment (muhammad-tanvir-1211) [inline benchmark/syclblas/blas3/symm.cpp:124]: Could you double check these calculations based on the file that I have shared with you?

File: cmake/CmakeFunctionHelper.cmake [hunk 12]
```diff
@@ -842,7 +848,10 @@ else() # default cpu backend
       add_gemm_configuration(
         "${data}"  64 "false" "false" "false"
         64 8 8 8 8 1 1 1 1 1 1 1 1 1 float float "no_local" "standard" "partial" 1 "strided" "false" "false")
-    endif()
+      add_gemm_configuration(
```
Comment (muhammad-tanvir-1211) [inline cmake/CmakeFunctionHelper.cmake:851]: I think you will need to update the `add_gemm_configuration` calls for all target hardwares to support `symm_a` and `symm_b` template parameters for those hardware.

**Review by muhammad-tanvir-1211 — APPROVED** [review]

**Review by muhammad-tanvir-1211 — CHANGES_REQUESTED** [review]
Accidentally approved the PR. I still need to go through the kernel to understand the implementation.

File: src/operations/blas3/gemm_local.hpp [hunk 13]
```diff
@@ -542,27 +582,42 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, TileType,
...
+        item_id, row_b, col_b, B, ldb, sB,
         [&](index_t ir, index_t cr)
             SYCL_BLAS_ALWAYS_INLINE { return cr < k - ir; },
-        [&](index_t, index_t cc) SYCL_BLAS_ALWAYS_INLINE { return cc < n; });
+        [&](index_t, index_t cc) SYCL_BLAS_ALWAYS_INLINE { return cc < n; },
+        [&](index_t row, index_t col) SYCL_BLAS_ALWAYS_INLINE { 
+          // If B is symmetric, the valid values are expected on the upper
+          // triangle unless B is transposed, in which case valid values will
+          // be on the lower side.
+          return symm_b ? trans_b ? row >= col : row <= col : true;
```
Comment (muhammad-tanvir-1211) [inline src/operations/blas3/gemm_local.hpp:619]: Same for this `lambda` here.

File: cmake/CmakeFunctionHelper.cmake (hunk 12 above)
Comment (aacostadiaz) [reply cmake/CmakeFunctionHelper.cmake:851]: I have added that in the loop above. Similar of what we do for `trans_a` and `trans_b`.

**Review by muhammad-tanvir-1211 — APPROVED** [review]

File: benchmark/syclblas/blas3/symm.cpp [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    // integer overflows for n_fl_ops and bytes_processed
+    double m_d = static_cast<double>(m);
+    double n_d = static_cast<double>(n);
+    double k_d = static_cast<double>(k);
+
+    state.counters["m"] = m_d;
+    state.counters["k"] = k_d;
+    state.counters["n"] = n_d;
+
+    double mem_read = (k_d * (k_d + 1) / 2) + 2 * (m_d * n_d);
```
Comment (s-Nick) [inline benchmark/syclblas/blas3/symm.cpp:115]: Thank you for you patch @aacostadiaz!
The counters in this benchmark differ slightly from `clblast` ones.

Comment (aacostadiaz) [reply benchmark/syclblas/blas3/symm.cpp]: Nice catch! Thanks. I updated the counters in the clblast benchmark but forgot this one.

**Review by s-Nick — APPROVED** [review]

**Review by OuadiElfarouki — APPROVED** [review]

**Pull request merged**